mlflow/tensorflow/__init__.py

```diff
@@ -887,6 +968,110 @@ def _log_early_stop_callback_metrics(callback, history, metrics_logger):
         if metric_key is not None:
             metrics_logger.record_metrics(restored_metrics, stopped_epoch + 1)
 
+    def _log_keras_model_with_warning(history, args):
+        try:
+            if log_models:
+                import types
+
+                def _get_input_data_slice():
+                    input_training_data_original = args[0]
+                    input_example_slice = None
+
+                    def _extract_from_sequence_or_generator():
+                        input_training_data = input_training_data_original
+                        if isinstance(input_training_data_original, types.GeneratorType):
+                            input_training_data = next(input_training_data_original)[0]
+                        elif isinstance(
+                            input_training_data_original, tensorflow.keras.utils.Sequence
+                        ):
+                            input_training_data = input_training_data_original[:][0]
+                        return input_training_data
+
+                    input_training_data = _extract_from_sequence_or_generator()
```

Comment:
Nit: I don't think we need to define a separate function here since it only references nonlocal state and is only ever called once. The following is easier to maintain:

```suggestion
                    input_training_data = args[0]
                    input_example_slice = None

                      if isinstance(input_training_data, types.GeneratorType):
                          input_training_data = next(input_training_data)[0]
                      elif isinstance(
                          input_training_data, tensorflow.keras.utils.Sequence
                      ):
                          input_training_data = input_training_data[:][0]
```

Reply:
done. thx